src/wallet/asyncrpcoperation_mergetoaddress.cpp

```diff
@@ -183,778 +113,87 @@ void AsyncRPCOperation_mergetoaddress::main()
-    // change upon arrival of new blocks which contain joinsplit transactions.  This is likely
-    // to happen as creating a chained joinsplit transaction can take longer than the block interval.
-    {
-        LOCK2(cs_main, pwalletMain->cs_wallet);
-        for (auto t : sproutNoteInputs_) {
-            JSOutPoint jso = std::get<0>(t);
-            std::vector<JSOutPoint> vOutPoints = {jso};
-            uint256 inputAnchor;
-            std::vector<std::optional<SproutWitness>> vInputWitnesses;
-            if (!pwalletMain->GetSproutNoteWitnesses(vOutPoints, nAnchorConfirmations, vInputWitnesses, inputAnchor)) {
-                // This error should not appear once we're nAnchorConfirmations blocks past
-                // Sprout activation.
-                throw JSONRPCError(RPC_WALLET_ERROR, "Insufficient Sprout witnesses.");
-            };
-            jsopWitnessAnchorMap[jso.ToString()] = MergeToAddressWitnessAnchorData{vInputWitnesses[0], inputAnchor};
-        }
-    }
-
-    /**
-     * SCENARIO #3
-     *
-     * zaddrs -> zaddr
-     * taddrs ->
-     *
-     * zaddrs ->
-     * taddrs -> taddr
-     *
-     * Send to zaddr by chaining JoinSplits together and immediately consuming any change
-     * Send to taddr by creating dummy z outputs and accumulating value in a change note
-     * which is used to set vpub_new in the last chained joinsplit.
-     */
-    UniValue obj(UniValue::VOBJ);
-    CAmount jsChange = 0;          // this is updated after each joinsplit
-    int changeOutputIndex = -1;    // this is updated after each joinsplit if jsChange > 0
-    bool vpubOldProcessed = false; // updated when vpub_old for taddr inputs is set in first joinsplit
-    bool vpubNewProcessed = false; // updated when vpub_new for miner fee and taddr outputs is set in last joinsplit
-
-    // At this point, we are guaranteed to have at least one input note.
-    // Use address of first input note as the temporary change address.
-    SproutSpendingKey changeKey = std::get<3>(zInputsDeque.front());
-    SproutPaymentAddress changeAddress = changeKey.address();
-
-    CAmount vpubOldTarget = 0;
-    CAmount vpubNewTarget = 0;
-    if (isToTaddr_) {
-        vpubNewTarget = z_inputs_total;
-    } else {
-        if (utxoInputs_.empty()) {
-            vpubNewTarget = minersFee;
+    try {
+        const auto& spendable = effects.GetSpendable();
+        const auto& payments = effects.GetPayments();
+        spendable.LogInputs(id);
+
+        bool sendsToShielded = false;
+        for (const auto& resolvedPayment : payments.GetResolvedPayments()) {
+            sendsToShielded = sendsToShielded || examine(resolvedPayment.address, match {
+                [](const CKeyID&) { return true; },
+                [](const CScriptID&) { return true; },
+                [](const libzcash::SaplingPaymentAddress&) { return false; },
+                [](const libzcash::OrchardRawAddress&) { return false; },
+            });
```

Comment:
Argh, I had originally written this for the `isToTaddr_` that existed before, then realized that we really want the opposite, because `isToTaddr` only works for the single-output case (and I expect this code to move in the near future to be shared across all operations). And I forgot to switch the `examine` when I inverted it 🤦🏼